sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala

```diff
@@ -79,6 +80,24 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
     udaf
   }
 
+  /**
+   * Registers a typed [[Aggregator]] for use with untyped Data Frames
+   *
+   * @param name the name to register under
+   * @param agg the typed Aggregator
+   * @return a UserDefinedAggregator that can be used as an aggregating UDF
+   *
+   * @since 3.0.0
+   */
+  def registerAggregator[IN: TypeTag, BUF, OUT](
```

Comment:
Another idea: instead of adding a new method here, we can create a new method `def udaf` in `org.apache.spark.sql.functions`.

There are already many `udf` methods there, which take type tag.

e.g.
```
def udaf[IN: TypeTag, BUF, OUT](agg: Aggregator[IN, BUF, OUT]): UserDefinedFunction = {
  ...
}
```

For java friendly, we can follow the same strategy in `Dateset`, which add an overload with encoder parameter:
```
def udaf[IN, BUF, OUT](agg: Aggregator[IN, BUF, OUT], inputEncoder: Encoder[IN]): UserDefinedFunction = {
  ...
}
```

Reply:
and users can register this udaf with the existing method `def register(name: String, udf: UserDefinedFunction)` defined in this file.